app-connector/src/server/connection.ts

```diff
@@ -0,0 +1,194 @@
+    })
+}
+
+async function callInfoUrl(infoUrl: any, crt: any, privateKey: any, insecure: any) {
+    LOGGER.debug("Calling info URL '%s'", infoUrl)
+
+    return request.get({
+        uri: infoUrl,
+        json: true,
+        agentOptions: {
+            cert: crt,
+            key: privateKey
+        },
+        rejectUnauthorized: !insecure,
+        resolveWithFullResponse: true
+    }).then((response: any) => {
+        if (response.statusCode !== 200) {
+            throw new Error("Calling Info URL failed with status '" + response.statusCode + "' and body '" + JSON.stringify(response.body, null, 2) + "'")
+        }
+        LOGGER.debug("Got following Info URL returned: %s", JSON.stringify(response.body, null, 2))
+        return response.body
+    })
+}
+function generateCSR(subject: any) {
+    LOGGER.debug("Creating CSR using subject %s", subject)
+    var pk = forge.pki.privateKeyFromPem(privateKey())
+    var publickey = forge.pki.setRsaPublicKey(pk.n, pk.e)
+
+    // create a certification request (CSR)
+    var csr = forge.pki.createCertificationRequest()
+    csr.publicKey = publickey
+
+    csr.setSubject(parseSubjectToJsonArray(subject))
+    csr.sign(pk)
+    LOGGER.debug("Created csr using subject %s", subject)
+    return forge.pki.certificationRequestToPem(csr)
+}
+
+function parseSubjectToJsonArray(subject: any) {
+    var subjectsArray: any = []
+    subject.split(",").map((el: any) => {
+        const val = el.split("=")
+        subjectsArray.push({
+            shortName: val[0],
+            value: val[1]
+        })
+    })
+
+    return subjectsArray;
+}
+
+function certificate() {
+    if (fs.existsSync(crtFile)) {
+        LOGGER.info("Found existing certificate: %s", crtFile)
+        return fs.readFileSync(crtFile, "utf-8")
+    }
+    return "";
+}
+
+function privateKey() {
+    if (fs.existsSync(privateKeyFile)) {
+        LOGGER.info("Found existing function key: %s", privateKeyFile)
+        return fs.readFileSync(privateKeyFile, "utf-8");
+    } else {
+        return generateprivateKey(privateKeyFile);
+    }
+}
+
+function established() {
+    if (!connection) {
+        connection = info();
+    }
+    return connection && connection.metadataUrl
+}
+
+function generateprivateKey(filePath: any) {
+    LOGGER.debug("Generating new function key: %s", filePath)
+    var keys = forge.pki.rsa.generateKeyPair(2048)
+    const key = forge.pki.privateKeyToPem(keys.privateKey)
+    fs.writeFileSync(filePath, key)
+    LOGGER.info("Generated new function key: %s", filePath)
+    return key
+}
+
+function info() {
+    if (fs.existsSync(connFile)) {
+        LOGGER.info("Found existing connection info: %s", connFile)
+        connection = JSON.parse(fs.readFileSync(connFile, "utf-8"))
+        return connection;
+    }
+    return null;
+}
+
+async function connect(tokenUrl: string, insecure: boolean = false, nodePort: any = null) {
+    if (!fs.existsSync(keysDirectory)) {

```

Comment:
Let's have an init method where we restore connection status from files. That will be called from app.ts once initially.

On connect, we write once all files, on disconnect we delete the files. Thats it, no other places where files are used, everything else is based on memory.
Initialization could have an option to disable persistence fully.

Reply:
done